scipy/interpolate/_aaa.py

```diff
@@ -0,0 +1,363 @@
+            # Interpolate at supp pts with `wj !=0`
+            D_inf = np.isinf(D) | np.isnan(D)
+            D[D_inf] = 1
+            N[D_inf] = f[D_inf]
+            R = N / D
+
+            # Check if converged
+            max_err = np.linalg.norm(f - R, ord=np.inf)
+            errvec[m] = max_err
+            if max_err <= atol:
+                break
+
+        if m == max_terms - 1:
+            raise RuntimeError(f"AAA failed to converge within {max_terms} iterations.")
+
+        # Trim off unused array allocation
+        zj = zj[: m + 1]
+        fj = fj[: m + 1]
+        C = C[:, : m + 1]
+        A = A[:, : m + 1]
+        errvec = errvec[: m + 1]
+
+        # Remove support points with zero weight
+        i_non_zero = wj != 0
+        zj = zj[i_non_zero]
+        wj = wj[i_non_zero]
+        fj = fj[i_non_zero]
+
+        self.support_points = zj
+        self.data_values = fj
+        self.weights = wj
+        self.errors = errvec
+
+        # Only compute poles, residues, and zeros when required
+        self._poles = None
+        self._residues = None
+        self._zeros = None
+
+    def __call__(self, z):
+        """Evaluate the rational approximation at given values.
+
+        Parameters
+        ----------
+        z : array_like
+            Input values.
+        """
+        # evaluate rational function in barycentric form.
+        z = np.asarray(z)
+        zv = np.ravel(z)
+
+        # Cauchy matrix
+        with np.errstate(invalid="ignore", divide="ignore"):
+            CC = 1 / np.subtract.outer(zv, self.support_points)
+        # Vector of values
+        with np.errstate(invalid="ignore"):
+            r = CC @ (self.weights * self.data_values) / (CC @ self.weights)
+
+        # Deal with input inf: `r(inf) = lim r(z) = sum(w*f) / sum(w)`
+        r[np.isinf(zv)] = np.sum(self.weights * self.data_values) / np.sum(self.weights)
+
+        # Deal with NaN
+        ii = np.flatnonzero(np.isnan(r))
+        for jj in ii:
+            if np.isnan(zv[jj]) or not np.any(zv[jj] == self.support_points):
+                # r(NaN) = NaN is fine.
+                # The second case may happen if `r(zv[ii]) = 0/0` at some point.
+                pass
+            else:
+                # Clean up values `NaN = inf/inf` at support points.
+                # Find the corresponding node and set entry to correct value:
+                r[jj] = self.data_values[zv[jj] == self.support_points].squeeze()
+
+        return np.reshape(r, z.shape)
+
+    @property
+    def poles(self):
+        if self._poles is None:
+            # Compute poles via generalized eigenvalue problem
+            m = self.weights.size
+            B = np.eye(m + 1)
+            B[0, 0] = 0
+
+            E = np.zeros_like(B, dtype=np.complex128)
+            E[0, 1:] = self.weights
+            E[1:, 0] = 1
+            np.fill_diagonal(E[1:, 1:], self.support_points)
+
+            pol = scipy.linalg.eigvals(E, B)
```

Comment:
Do we want to use NumPy eigvals or SciPy eigvals? I'd normally default to NumPy eigvals, but perhaps you have a good reason for SciPy?

Reply:
NumPy eigvals does not support generalised eigenvalue problems so it is necessary to use scipy here